#### Data Analysis with IMDB Movie Data

IMDB Movie Dataset là một bộ dữ liệu đánh giá phim, dùng để phân tích mức độ quan tâm của phim theo một số tiêu chí như: đạo diễn, diễn viên, tên phim,... nhằm đưa ra những góc nhìn hỗ tr dự đoán trong tương lai.


##### Các bước cần thực hiện trong bài toán:
1. Read Data
2. View the Data
3. Understand some basic information about the data
4. Data Selection - Indexing and Slicing data
5. Data Selection - Based on Conditional Filtering
6. Group by operations
7. Sorting operation
8. View missing values
9. Deal with missing values - deleting
10. Deal with missing values - Filling
11. Apply()

1. Import Libraries and Load Dataset

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

dataset_path = '../../../data/module3/week1/IMDB-Movie-Data.csv'

data = pd.read_csv(dataset_path)

Ta chọn cột Title làm cột chỉ mục (Cột chỉ mục là cột mà ở đó ko chưa giá trị trùng lặp)

In [2]:
data_indexed = pd.read_csv(dataset_path, index_col = "Title")

2. View the data

Sử dụng head để hiển thị 5 hàng đầu tiên của bảng dữ liệu 

In [3]:
data.head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0  
2     7.3  157606              138.12       62.0  
3     7.2   60545              270.32       59.0  
4     6.2  393727              325.02       40.0

3. Understand some basic information about the data

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


In [5]:
data.describe()

Rank         Year  Runtime (Minutes)       Rating         Votes  \
count  1000.000000  1000.000000        1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000         113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962          18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000          66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000         100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000         111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000         123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000         191.000000     9.000000  1.791916e+06   

       Revenue (Millions)   Metascore  
count          872.000000  936.000000  
mean            82.956376   58.985043  
std            103.253540   17.194757  
min              0.000000   11.000000  
25%             13.270000   47.000000  
50%             47.985000   59.500000  
75%            113.715000   72.000000  
max            936.630000  100.000000

Ở đây ta có thể thấy:
- Giá trị min và max của Year, tức dataset chưa các bộ phim từ 2006 tới 2016
- rating trung bình cho các bộ phim là 6.7, thấp nhất là 1.9, cao nhất là 9.0
- Doanh thu cao nhất đạt được là 936.6 triệu dollar

4. Data selection - Indexing and Slicing Data 

Ở đây, ta sẽ tách một số cột trong data sử dụng kỹ thuật Indexing, để tách cột thành series, ta thực hiện:

In [6]:
genre = data['Genre']
genre

0       Action,Adventure,Sci-Fi
1      Adventure,Mystery,Sci-Fi
2               Horror,Thriller
3       Animation,Comedy,Family
4      Action,Adventure,Fantasy
                 ...           
995         Crime,Drama,Mystery
996                      Horror
997         Drama,Music,Romance
998            Adventure,Comedy
999       Comedy,Family,Fantasy
Name: Genre, Length: 1000, dtype: object

Để tách cột thành DataFrame, ta thực hiện như sau:

In [7]:
data[['Genre']]

Genre
0     Action,Adventure,Sci-Fi
1    Adventure,Mystery,Sci-Fi
2             Horror,Thriller
3     Animation,Comedy,Family
4    Action,Adventure,Fantasy
..                        ...
995       Crime,Drama,Mystery
996                    Horror
997       Drama,Music,Romance
998          Adventure,Comedy
999     Comedy,Family,Fantasy

[1000 rows x 1 columns]

Ta có thể chọn và tách cùng một lúc nhiều cột với nhau, tạo thành một DataFrame mới:

In [8]:
some_col = data[['Title', 'Genre', 'Actors', 'Director', 'Rating']]
some_col

Title                     Genre  \
0    Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1                 Prometheus  Adventure,Mystery,Sci-Fi   
2                      Split           Horror,Thriller   
3                       Sing   Animation,Comedy,Family   
4              Suicide Squad  Action,Adventure,Fantasy   
..                       ...                       ...   
995     Secret in Their Eyes       Crime,Drama,Mystery   
996          Hostel: Part II                    Horror   
997   Step Up 2: The Streets       Drama,Music,Romance   
998             Search Party          Adventure,Comedy   
999               Nine Lives     Comedy,Family,Fantasy   

                                                Actors              Director  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...            James Gunn   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...          Ridley Scott   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...    M. Night Shyamalan   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  Christophe Lourdelet   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...            David Ayer   
..                                                 ...                   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...             Billy Ray   
996  Lauren German, Heather Matarazzo, Bijou Philli...              Eli Roth   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...            Jon M. Chu   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...        Scot Armstrong   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...      Barry Sonnenfeld   

     Rating  
0       8.1  
1       7.0  
2       7.3  
3       7.2  
4       6.2  
..      ...  
995     6.2  
996     5.5  
997     6.2  
998     5.6  
999     5.3  

[1000 rows x 5 columns]

Đối với việc Tách hàng, ta có thể tách ra một số lượng hàng nhất định, từ chỉ mục X đến chỉ mục Y trong bảng dữ liệu, gọi là slicing. Ví dụ, để tách các hàng thứ 10 -> thứ 15, ta làm như sao:

In [9]:
data.iloc[10:15][['Title', 'Rating', 'Revenue (Millions)']]

Title  Rating  Revenue (Millions)
10  Fantastic Beasts and Where to Find Them     7.5              234.02
11                           Hidden Figures     7.8              169.27
12                                Rogue One     7.9              532.17
13                                    Moana     7.7              248.75
14                                 Colossal     6.4                2.87

5. Data Selection - Based on Conditional Filtering:

Ta có lấy hàng trong bảng dữ liệu dựa trên một số điều kiện cần tuân theo. Ví dụ, ta mong muốn lấy các bộ phim từ năm 2010-2015, với rating nhỏ hơn 6.0 nhưng lại có doanh thu thuộc top 5% trên toàn bộ dataset. Theo đó, ta có thể triển khai code như sau:

In [11]:
data[((data['Year'] >= 2010) & (data['Year'] <= 2015))
     & (data['Rating'] < 6.0)
     & (data['Revenue (Millions)'] > data['Revenue (Millions)'].quantile(0.95))]

Rank                       Title                    Genre  \
941   942  The Twilight Saga: Eclipse  Adventure,Drama,Fantasy   

                                           Description     Director  \
941  As a string of mysterious killings grips Seatt...  David Slade   

                                                Actors  Year  \
941  Kristen Stewart, Robert Pattinson, Taylor Laut...  2010   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
941                124     4.9  192740              300.52       58.0

6. Groupby Operations: Groupby là một phép gom nhóm dữ liệu dựa trên một hoặc nhiu biến(Ở đây là cột dữ liệu trong bảng). Ví dụ, Ta có thể tìm số rating trung bình mà các đạo diễn đạt được bằng cách gom nhóm các chỉ số Rating bộ phim theo Director.

In [12]:
data.groupby('Director')[['Rating']].mean().head()

Rating
Director                   
Aamir Khan              8.5
Abdellatif Kechiche     7.8
Adam Leon               6.5
Adam McKay              7.0
Adam Shankman           6.3

7. Sorting Operations:

Sorting cho phép ta sắp xếp các hàng trong bảng dữ liệu theo thứ tự tăng/ giảm dần dựa theo giá tị của cột nào đó trong bảng dữ liệu. Ví dụ, dựa trên kết quả groupby phần trước, ta có thể tìm top 5 đjao diễn đạt số rating trung bình cao nhất như sau:


In [13]:
data.groupby('Director')[['Rating']].mean().sort_values(['Rating'], ascending = False).head()

Rating
Director                 
Nitesh Tiwari        8.80
Christopher Nolan    8.68
Olivier Nakache      8.60
Makoto Shinkai       8.60
Aamir Khan           8.50

8. View Missing Values: Các bộ dữ liệu thường sẽ xuất hiện tình trạng bị giá trị rỗng (missing value) trong một vài trường thông tin của một số mẫu dữ liệu. Khi xử lý dữ liệu, ta cần khắc phục vấn đề này. Vì vậy, việc đầu tiên  ta cần kiểm tra xem vị trí bị mất mát dữ liệu theo cách sau:

In [14]:
data.isnull().sum()

Rank                    0
Title                   0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
dtype: int64

Ở đây ta thy Revenue(Millions) và Metascore là 2 cột có chứa dữ liệu null. Để xử lý vấn đề mất mát dữ liệu, có hai phương án chính: hoặc thế các vùng trống bằng một giá trị nào đó hoặc loại bỏ chúng.

9. Deal with missing values - Deleting

Đối với phương án loại bỏ, ta có thể loại bỏ toàn ộ cột chứa nhiều giá trị null (nếu có thể ) hoặc chỉ loại bỏ các hàng chứa giá trị null. Đối với xóa cột, ta thực hiện:

In [16]:
data.drop('Metascore', axis = 1, inplace = True)

In [17]:
data.dropna()

Rank                     Title                     Genre  \
0       1   Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2                Prometheus  Adventure,Mystery,Sci-Fi   
2       3                     Split           Horror,Thriller   
3       4                      Sing   Animation,Comedy,Family   
4       5             Suicide Squad  Action,Adventure,Fantasy   
..    ...                       ...                       ...   
993   994  Resident Evil: Afterlife   Action,Adventure,Horror   
994   995                 Project X                    Comedy   
996   997           Hostel: Part II                    Horror   
997   998    Step Up 2: The Streets       Drama,Music,Romance   
999  1000                Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
993  While still out to destroy the evil Umbrella C...    Paul W.S. Anderson   
994  3 high school seniors throw a birthday party t...       Nima Nourizadeh   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
993  Milla Jovovich, Ali Larter, Wentworth Miller,K...  2010   
994  Thomas Mann, Oliver Cooper, Jonathan Daniel Br...  2012   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  
0                  121     8.1  757074              333.13  
1                  124     7.0  485820              126.46  
2                  117     7.3  157606              138.12  
3                  108     7.2   60545              270.32  
4                  123     6.2  393727              325.02  
..                 ...     ...     ...                 ...  
993                 97     5.9  140900               60.13  
994                 88     6.7  164088               54.72  
996                 94     5.5   73152               17.54  
997                 98     6.2   70699               58.01  
999                 87     5.3   12435               19.64  

[872 rows x 11 columns]

10. Dealing with missing value - Filling:

Đối với phương án thay thế giá trị mới vào các ô trống, ta có thể sử dụng các giá trị mean, median... của cocojt dữ liệu tương ứng để thay thế ( việc chọn giá trị để thay th còn tùy thuộc vào tính cht của bộ dữ liệu, bài toán đang giải quyết...). Ví dụ, có một vài hàng có Revenue mang giá trị null, ta có thể gán cho nó giá trị trung bình như sau:


In [21]:
revenue_mean = data_indexed['Revenue (Millions)'].mean()
print("The mean revenue is: ", revenue_mean)

# we can fill the null values with thí mean revenue
data_indexed.fillna({'Revenue (Millions)': revenue_mean}, inplace = True)

The mean revenue is:  82.95637614678898


11. apply() functiions:

Apply function được dùng khi ta muốn thực thi một hàm nào đó lên các hàng trong bảng dữ liệu. Sau khi thực thi, kết quả trả về từ hàm chính là giá trị mới của hàng tương ứng. Ví dụ, ta muốn phân loại phim theo ba mức độ ['Good','Average','Bad'] dựa trên Rating, ta có thể định nghĩa một hàm để làm điều này và apply nó lên DataFrame:

In [23]:
def rating_group(rating):
	if rating >= 7.5:
		return 'Good'
	elif rating >= 6.0:
		return 'Average'
	else:
		return 'Bad'



In [24]:
data['Rating_category'] = data['Rating'].apply(rating_group)
data[['Title', 'Director', 'Rating', 'Rating_category']].head(5)

Title              Director  Rating Rating_category
0  Guardians of the Galaxy            James Gunn     8.1            Good
1               Prometheus          Ridley Scott     7.0         Average
2                    Split    M. Night Shyamalan     7.3         Average
3                     Sing  Christophe Lourdelet     7.2         Average
4            Suicide Squad            David Ayer     6.2         Average